staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -529,18 +529,11 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return nil, errors.NewServiceUnavailable(err.Error())
 	}
 
-	// determine the namespace and name scope of the watch, first from the request, secondarily from the field selector
-	scope := namespacedName{}
-	if requestNamespace, ok := request.NamespaceFrom(ctx); ok && len(requestNamespace) > 0 {
-		scope.namespace = requestNamespace
-	} else if selectorNamespace, ok := pred.Field.RequiresExactMatch("metadata.namespace"); ok {
-		scope.namespace = selectorNamespace
-	}
-	if requestInfo, ok := request.RequestInfoFrom(ctx); ok && requestInfo != nil && len(requestInfo.Name) > 0 {
-		scope.name = requestInfo.Name
-	} else if selectorName, ok := pred.Field.RequiresExactMatch("metadata.name"); ok {
-		scope.name = selectorName
+	requestInfo, ok := request.RequestInfoFrom(ctx)
```

Comment:
Before I start fixing the unit tests I wanted to pick your brain on this change. 

The RequestInfo must always be present otherwise the chain wasn't wired correctly.
Do you remember why we had a custom logic for determining the request scope ?

Reply:
I can't remember it, but I vaguely recall  Jordan was pushing to not use RequestInfo.

@liggitt - do you remember this?

Reply:
The current code handles requests to a cluster-wide endpoint that sets a selector that scopes to a specific name or namespace:

* `/api/v1/pods?metadata.namespace=...`
* `/api/v1/pods?metadata.name=...`

Dropping the selector scoping makes those requests non-performant again.

Reply:
If there's no test that failed with this change, we have a test gap we need to close to make sure we don't lose this behavior :-/

Reply:
ok, I'd like to add this logic to `NewRequestInfo` for `watch` requests only, would that be okay with you  ? (https://github.com/kubernetes/kubernetes/blob/bc6c7fa91201348d010b638fbadf32007c0ac546/staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go#L117)

Reply:
`NewRequestInfo` already knows how to deal with `metadata.name`, it misses `metadata.namespace` case.

Reply:
https://github.com/kubernetes/kubernetes/pull/120490/files#diff-dde802c9d180be5b853213466f68312c6f403e8a22d30e20790d07aeec7352e9R393 is where I would like to use it.